runtime/src/accounts_db.rs

```diff
@@ -1958,51 +1963,56 @@ impl AccountsDb {
-                let lookup = if is_startup {
-                    self.accounts_index.get_account_read_entry_with_lock(
-                        pubkey,
-                        accounts_index_map_lock_ref.unwrap(),
-                    )
-                } else {
-                    self.accounts_index.get_account_read_entry(pubkey)
-                };
+        for (pubkey, stored_account) in &stored_accounts {
+            let lookup = if is_startup {
+                self.accounts_index
+                    .get_account_read_entry_with_lock(pubkey, accounts_index_map_lock_ref.unwrap())
+            } else {
+                self.accounts_index.get_account_read_entry(pubkey)
+            };
 
-                if let Some(locked_entry) = lookup {
-                    let is_alive = locked_entry.slot_list().iter().any(|(_slot, i)| {
-                        i.store_id == stored_account.store_id
-                            && i.offset == stored_account.account.offset
-                    });
-                    if !is_alive {
-                        // This pubkey was found in the storage, but no longer exists in the index.
-                        // It would have had a ref to the storage from the initial store, but it will
-                        // not exist in the re-written slot. Unref it to keep the index consistent with
-                        // rewriting the storage entries.
-                        locked_entry.unref()
-                    } else {
-                        alive_total += stored_account.account_size as u64;
-                    }
-                    is_alive
+            if let Some(locked_entry) = lookup {
+                let is_alive = locked_entry.slot_list().iter().any(|(_slot, i)| {
+                    i.store_id == stored_account.store_id
+                        && i.offset == stored_account.account.offset
+                });
+                if !is_alive {
+                    // This pubkey was found in the storage, but no longer exists in the index.
+                    // It would have had a ref to the storage from the initial store, but it will
+                    // not exist in the re-written slot. Unref it to keep the index consistent with
+                    // rewriting the storage entries.
+                    unrefed_pubkeys.push(pubkey);
+                    locked_entry.unref();
                 } else {
-                    false
+                    alive_total += stored_account.account_size as u64;
+                    alive_accounts.push((pubkey, stored_account));
                 }
-            })
-            .collect();
+            }
+        }
         drop(accounts_index_map_lock);
         index_read_elapsed.stop();
         let aligned_total: u64 = self.page_align(alive_total);
 
+        // If not saving memory maps or more than 1 page of data, then skip the shrink
+        if num_stores == 1 && aligned_total + PAGE_SIZE >= original_bytes {
```

Comment:
I think it's also a good idea to extract this into a reusable `is_bytes_saved_sufficient_for_shrink()` that can be added to the check here where we add to the `shrink_candidates`: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_db.rs#L4592-L4607.

Otherwise, we'll continually add to the shrink candidates as we delete accounts from the same slot, even if they continually fail this new `aligned_total + PAGE_SIZE >= original_bytes` check

Reply:
Also, just out of curiosity, do we know the snapshot sizes before and after?

Theoretically if every slot had an extra `PAGE_SIZE = 4096` bytes, we might get an extra `432000 slots * 496 bytes ~= 2GB` raw extra size before compression, but wonder if the results reflect this.

Reply:
I ran this for a while on mainnet. snapshot sizes didn't seem to be affected:

```
Every 2.0s: ls -alh mainnet-beta/validator-ledger/snapshot-*

-rw-rw-r-- 1 stephen_solana_com stephen_solana_com 6.3G May  4 15:00 mainnet-beta/validator-ledger/snapshot-76776676-EcV4U9gqqdXuhqgP69WaLhWmoyDQcJYxM1ZE5YuSspdF.tar.zst
-rw-rw-r-- 1 stephen_solana_com stephen_solana_com 6.6G May 27 15:13 mainnet-beta/validator-ledger/snapshot-80249526-9NwTfazvvgmdb4Wr8VnAhC1kEWqnU9APZjfpmrktjpRS.tar.zst
-rw-rw-r-- 1 stephen_solana_com stephen_solana_com 6.6G May 27 15:18 mainnet-beta/validator-ledger/snapshot-80249917-2WYkAHaKCYFEyCtA8gBF4Qxh6iNbrcxRbXUrqUQSvPai.tar.zst
```

Reply:
Left side is this change, right side master starting with that huge spike at startup.

<img width="1803" alt="bytes-written" src="https://user-images.githubusercontent.com/1024962/119875574-33e1d500-bf27-11eb-8ffc-d504c4062c72.png">